### ISSUE TITLE
Semanticsdb is missing synthetics information about implicits in for comprehensions

### ISSUE THREAD

**Issue by fikovnik** [scalameta/scalameta#1133]
I have the following file (`Input2.scala`):

```scala
import scala.concurrent.Future

object Input2 {
  import scala.concurrent.ExecutionContext.Implicits.global

  val f1 = Future { 1 }
  val f2 = Future { 2 }

  f1.flatMap(v1 => f2.map(v2 => v1 + v2))
}
```
The synthetics part of the corresponding semanticsdb correctly shows 4 occurrences of the global execution context being inserted at the right places:

```
scala/fix/Input2.scala
195 -- *(scala.concurrent.ExecutionContext.Implicits.global)
269 -- *(scala.concurrent.ExecutionContext.Implicits.global)
214 -- *.apply[Int]
268 -- *(scala.concurrent.ExecutionContext.Implicits.global)
226 -- *(scala.concurrent.ExecutionContext.Implicits.global)
183 -- *.apply[Int]
253 -- *[Int]
240 -- *[Int]
```
However, if I rewrite the code above to (`Input1.scala`):
```scala
import scala.concurrent.Future

object Input2 {
  import scala.concurrent.ExecutionContext.Implicits.global

  val f1 = Future { 1 }
  val f2 = Future { 2 }

  for {
   v1 <- f1
   v2 <- f2
  } yield v1 + v2
}
```
I get only 3 occurrences:
```
scala/fix/Input1.scala
183 -- *.apply[Int]
277 -- *(scala.concurrent.ExecutionContext.Implicits.global)
214 -- *.apply[Int]
261 -- *.map[Int]
195 -- *(scala.concurrent.ExecutionContext.Implicits.global)
248 -- *.flatMap[Int]
226 -- *(scala.concurrent.ExecutionContext.Implicits.global)
```
In the desugaring of for comprehensions, one use of implicit is missing.

An example project (using scalafix to setup things quickly) is here:
https://www.dropbox.com/s/fsg8z86p9spe4n9/semdb-forcomp.tar.bz2?dl=0

Using scalameta 2.0.1, scala 2.12.3

**Comment by fikovnik** [comment]
I had a quick look at the scalameta sources and the problem is that for both `flatMap` and `map` the `ApplyToImplicitArgs` has the same end position. Since the storage is a map, they get overwritten.

**Comment by olafurpg** [comment]
Thank you for reporting! We use `Inferred` to merge multiple synthetics with the same position, see https://github.com/scalameta/scalameta/blob/7b58f0e0a54b4d925668dc8f721881cb40da419f/scalameta/semanticdb-scalac/src/main/scala/scala/meta/internal/semanticdb/Inferred.scala#L32-L37

What do you think would be sensible locations for the implicit synthetics? They should clearly not be the same, but I'm not sure what position makes more sense. Merging `*.flatMap` and `*(ExectuionContext.Implicits)` into `*.flatMap(Implicits)` is not correct.

**Comment by olafurpg** [comment]
Making `Inferred.args` a list is one possible solution, but that would produce non-sensical synthetics like `(Implicit)(Implicit)`. Do you have an idea how we could distinguish the inferred `Implicit`s that have the same position? It would be nice if we could someone know which one is attached to which .flatmap/.map

**Comment by fikovnik** [comment]
Indeed, it would be really good to know which one is which.

I'm not sure if I completely get the meaning of `_star_`, but it looks like a placeholder. In the `Input2.scala` (from above) you can basically replace the star by whatever is on the given position and you get out a valid desugared code. I guess that is the purpose, right?

For the for comprehension, this does not seem to be working well, since it is a more complicated rewrite. In this case wouldn't it make sense to actually merge the whole thing? For example:

```scala
for (v1 <- f1; v2 <- f2) yield v1+v2
^                              ^   ^
0                              31  35
```
will produce:
```
[0..35) f1.flatMap[Int](((v1: Int) => f2.map[Int](((v2: Int) => v1.+(v2)))(_root_.scala.concurrent.ExecutionContext.Implicits.global)))(_root_.scala.concurrent.ExecutionContext.Implicits.global)
```
or
```
[0..31) <empty string signalling removal>
[31..35) f1.flatMap[Int](((v1: Int) => f2.map[Int](((v2: Int) => *))(_root_.scala.concurrent.ExecutionContext.Implicits.global)))(_root_.scala.concurrent.ExecutionContext.Implicits.global)
```
and the names in both cases will contain all the corresponding resolved names (`flatMap, map, Int, global, ...`). 

Perhaps this will break a lot of other things, but for our use case it will be perferct :-)

**Comment by heathermiller** [comment]
Gentle nudge 😉😉 @olafurpg? Turns out this important for the work that @fikovnik et al are doing analyzing implicits code. I think @fikovnik would be up to make a PR, but might need to be pointed in the right direction...

**Comment by olafurpg** [comment]
Sorry, totally forgot this, thanks for the reminder @heathermiller 

I think the solution proposed by @fikovnik is the best, we can make `args: List[AttributedSynthetic ]`. Would that solve your use-case @fikovnik ? If so, would you be interested in making a PR doing the change? I suspect the diff would be small, I'm curious to know if our tests will be affected.

Synthetics for for-comprehensions are a best-effort for now anyways, so I believe it's fair to over-persist data instead of under-persist.

> replace the star by whatever is on the given position and you get out a valid desugared code. I guess that is the purpose, right?

Correct. That is at least the idea for implicit conversions in particular. I agree this doesn't seem to scale well for for-comprehensions.

**Comment by fikovnik** [comment]
@olafurpg - yes I'll be happy tot try to make a PR.

Making it a `List[AttributedSynthetics]` will work for us, however, it will break the assumption with `_star_`. Another option would be to turn:

```scala
for (v1 <- f1; v2 <- f2) yield v1+v2
```
into
```
[0..35) f1.flatMap[Int](((v1: Int) => f2.map[Int](((v2: Int) => v1.+(v2)))(_root_.scala.concurrent.ExecutionContext.Implicits.global)))(_root_.scala.concurrent.ExecutionContext.Implicits.global)
```
or
```scala
[0..31) <empty string signalling removal>
[31..35) f1.flatMap[Int](((v1: Int) => f2.map[Int](((v2: Int) => *))(_root_.scala.concurrent.ExecutionContext.Implicits.global)))(_root_.scala.concurrent.ExecutionContext.Implicits.global)
```
which will preserve it.

Which one would you prefer?

**Comment by olafurpg** [comment]
@fikovnik I don't like the idea of persisting the full desugaring of for-comprehensions in synthetics since it would bloat the semanticdb.

I like where you are going in the end by using star to replace the non-synthetic argument. Can we take that idea further by changing the position of the implicit argument to follow flatMap/map/foreach? The output would be something like

```
[35] .flatMap[Int](*)(ExecutionContext)
[42] .map[Int](*)(ExecutionContext2)
```
even if the execution context is not part of

**Comment by olafurpg** [comment]
Fixed by #1142
